### PR TITLE
[TRTLLM-7192][feat] optimize MLA chunked prefill && support fp8 mla chunked prefill

### PR DESCRIPTION
**Pull request by jmydurant**
## Summary by CodeRabbit

- New Features
  - Enable saving softmax stats for FP8/E4M3 attention (formerly blocked).
  - Add configurable chunked prefill buffer batch size; exposed in Python APIs and used in execution.
  - Introduce global-offset–based chunked prefill with per-loop max length for more efficient KV loading.
  - Broaden support to SM90/SM120 for FP8 MLA, KV cache reuse, and chunked context.
  - Adjust workspace sizing to scale with chunked prefill buffer batch size.

- Tests
  - Expand FP8 coverage in accuracy tests and update H100 test lists to exercise FP8 chunked prefill paths.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/attention_backend/interface.py [hunk 1]
```diff
@@ -25,6 +25,7 @@ class AttentionRuntimeFeatures:
     cache_reuse: bool = False
     has_speculative_draft_tokens: bool = False
     chunk_size: int = 0  # this is the chunk size for MLA chunked prefill, it will split kv cache into chunks to save global memory.
+    chunk_buffer_batch_size: int = 4  # real chunk size for MLA chunked prefill is chunk_buffer_batch_size * chunk_size.
```
Comment (PerkzZheng) [inline tensorrt_llm/_torch/attention_backend/interface.py:28]: Can you explain more about why we add this parameter. What is difference compared to just setting larger chunk size ?

And if it helps with performance, have we collected any performance numbers ? thanks.

File: tensorrt_llm/_torch/attention_backend/interface.py [hunk 2]
```diff
@@ -630,3 +631,4 @@ class MLAParams:
     qk_nope_head_dim: int = 0
     v_head_dim: int = 0
     predicted_tokens_per_seq: int = 1
+    chunk_prefill_buffer_batch_size: int = 1
```
[the remark this answers is not in the document]
Comment (jmydurant) [reply tensorrt_llm/_torch/attention_backend/interface.py]: Done, all the params are rename to `chunked_prefill_buffer_batch_size`

File: tensorrt_llm/_torch/attention_backend/trtllm.py [hunk 3]
```diff
@@ -813,24 +820,49 @@ def pre_process_for_chunked_prefill(
...
             self.kv_cache_params.num_cached_tokens_per_seq,
             dtype=torch.int,
             device='cpu',
         )
+        remain_buffer_len = total_chunk_size
+        current_batch_idx = 0
+        max_chunk_len_per_loop.clear()
+        temp_max_chunk_len = 0
+        # cal chunked_seq_len
+        for b_id in range(num_contexts):
```
[the remark this answers is not in the document]
Comment (jmydurant) [reply tensorrt_llm/_torch/attention_backend/trtllm.py:835]: Yeah, just refer to the first comment.

File: tensorrt_llm/_torch/attention_backend/interface.py (hunk 1 above)
Comment (PerkzZheng) [reply tensorrt_llm/_torch/attention_backend/interface.py]: IIUC, you are trying to remove the paddings between the batches when splitting it into chunks for more efficient memory usage and computation. 

then, why we need two parameters here ? can it work with only chunk size = 128 and remove those paddings ?

File: tensorrt_llm/_torch/attention_backend/trtllm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (PerkzZheng) [reply tensorrt_llm/_torch/attention_backend/trtllm.py:835]: it is better to add the same comments to the codes. Thanks.

File: tensorrt_llm/_torch/attention_backend/interface.py (hunk 1 above)
Comment (jmydurant) [reply tensorrt_llm/_torch/attention_backend/interface.py]: Initially, the logic behind setting the chunk size was consistent with the batch scheduler, both based on max_num_tokens (a user-specifiable value). However, for performance considerations, particularly in long-text scenarios, our analysis suggests that allocating a buffer size 2 to 4 times this value is more reasonable. Therefore, we adopted this approach, using 4 times the user-specified max_num_tokens as the buffer size for chunked prefill.

**Review by PerkzZheng — APPROVED** [review]

**Review by QiJune — APPROVED** [review]

File: tensorrt_llm/_torch/attention_backend/interface.py [hunk 4]
```diff
@@ -630,3 +631,4 @@ class MLAParams:
     qk_nope_head_dim: int = 0
     v_head_dim: int = 0
     predicted_tokens_per_seq: int = 1
+    chunked_prefill_buffer_batch_size: int = 1
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/interface.py:634]: Why we use different default value from AttentionRuntimeFeatures's chunked_prefill_buffer_batch_size, which is 4?

File: tensorrt_llm/_torch/attention_backend/trtllm.py [hunk 5]
```diff
@@ -187,6 +188,7 @@ def plan(
         spec_decoding_packed_mask: Optional[torch.Tensor] = None,
         spec_decoding_generation_lengths: Optional[torch.Tensor] = None,
         attention_sinks: Optional[torch.Tensor] = None,
+        chunked_prefill_buffer_batch_size: int = 1,
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/trtllm.py:191]: Please add doc string for `chunked_prefill_buffer_batch_size`

File: tensorrt_llm/_torch/attention_backend/trtllm.py [hunk 6]
```diff
@@ -861,14 +893,17 @@ def prepare_context_mla_with_cached_kv(self,
             # currently we assume that the chunk size is the same as the max_num_tokens
             if self.runtime_features.chunked_prefill:
                 chunk_size = self.runtime_features.chunk_size
-                self.chunked_loop_num = (self.max_ctx_cached_token_len +
-                                         chunk_size - 1) // chunk_size
-                self.chunked_seq_len = torch.empty(
+                chunk_batch_size = self.runtime_features.chunked_prefill_buffer_batch_size
+                total_chunk_size = chunk_size * chunk_batch_size
+                self.chunked_loop_num = (self.num_ctx_cached_tokens +
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/trtllm.py:898]: Can we use `self.chunked_loop_num = math.ceil(self.num_ctx_cached_tokens/total_chunk_size)` instead?

File: tensorrt_llm/_torch/attention_backend/trtllm.py [hunk 7]
```diff
@@ -813,24 +820,49 @@ def pre_process_for_chunked_prefill(
...
             dtype=torch.int,
             device='cpu',
         )
+        remain_buffer_len = total_chunk_size
+        current_batch_idx = 0
+        max_chunk_len_per_loop.clear()
+        temp_max_chunk_len = 0
+        # cal chunked_seq_len
+        for b_id in range(num_contexts):
+            temp_cached_kv_len = cached_kv_lens[b_id].item()
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/trtllm.py:836]: We don't need to convert `num_cached_tokens_per_seq` to `cached_kv_lens` and access single item from it. Directly use `num_cached_tokens_per_seq[b_id]` is faster since that we avoid the pybind overhead. Can we remove `cached_kv_lens`?

File: tensorrt_llm/_torch/attention_backend/trtllm.py [hunk 8]
```diff
@@ -813,24 +820,49 @@ def pre_process_for_chunked_prefill(
...
+            while temp_cached_kv_len > 0:
+                used_buffer_len = min(remain_buffer_len, temp_cached_kv_len)
+                chunked_seq_len[current_batch_idx, b_id] = used_buffer_len
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/trtllm.py:839]: Can we use `batch_idx` instead of `b_id`?

File: tensorrt_llm/_torch/attention_backend/trtllm.py [hunk 9]
```diff
@@ -813,24 +820,49 @@ def pre_process_for_chunked_prefill(
...
+        total_chunk_size = chunk_size * chunk_batch_size
         cached_kv_lens = torch.tensor(
             self.kv_cache_params.num_cached_tokens_per_seq,
             dtype=torch.int,
             device='cpu',
         )
+        remain_buffer_len = total_chunk_size
+        current_batch_idx = 0
+        max_chunk_len_per_loop.clear()
+        temp_max_chunk_len = 0
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/trtllm.py:833]: Can we rename `temp_max_chunk_len/temp_cached_kv_len` to `max_chunk_len/cached_kv_len`? We don't need the `temp` prefix since no naming conflict.

File: tensorrt_llm/_torch/attention_backend/trtllm.py [hunk 10]
```diff
@@ -813,24 +820,49 @@ def pre_process_for_chunked_prefill(
...
+        # cal chunked_seq_len
+        for b_id in range(num_contexts):
+            temp_cached_kv_len = cached_kv_lens[b_id].item()
+            while temp_cached_kv_len > 0:
+                used_buffer_len = min(remain_buffer_len, temp_cached_kv_len)
+                chunked_seq_len[current_batch_idx, b_id] = used_buffer_len
+                temp_max_chunk_len = max(temp_max_chunk_len, used_buffer_len)
+                remain_buffer_len -= used_buffer_len
+                temp_cached_kv_len -= used_buffer_len
+                chunked_ld_global_offset[
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/trtllm.py:843]: Can we use `torch.cumsum` to calculate `chunked_ld_global_offset` from `chunked_seq_len` in a vectorized way?

File: tensorrt_llm/_torch/attention_backend/trtllm.py [hunk 11]
```diff
@@ -883,6 +918,17 @@ def prepare_context_mla_with_cached_kv(self,
                     device='cpu',
                     pin_memory=True,
                 )
+                self.chunked_ld_global_offset = torch.zeros(
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/trtllm.py:921]: Can we rename `chunked_ld_global_offset` to `chunked_global_offset` instead? The `ld` part is only necessary when used in the kernel code.

File: tensorrt_llm/_torch/attention_backend/trtllm.py (hunk 5 above)
Comment (jmydurant) [reply tensorrt_llm/_torch/attention_backend/trtllm.py:191]: Done.

File: tensorrt_llm/_torch/attention_backend/trtllm.py (hunk 7 above)
Comment (jmydurant) [reply tensorrt_llm/_torch/attention_backend/trtllm.py]: Done.

File: tensorrt_llm/_torch/attention_backend/trtllm.py (hunk 9 above)
Comment (jmydurant) [reply tensorrt_llm/_torch/attention_backend/trtllm.py]: Done.

File: tensorrt_llm/_torch/attention_backend/trtllm.py (hunk 11 above)
Comment (jmydurant) [reply tensorrt_llm/_torch/attention_backend/trtllm.py]: Done.

File: tensorrt_llm/_torch/attention_backend/interface.py (hunk 4 above)
Comment (yuxianq) [reply tensorrt_llm/_torch/attention_backend/interface.py:653]: Got it.

**Review by yuxianq — APPROVED** [review]
